[Doc] [autodiff] Add a section about customized gradient in autodiff. (#3054)
taichi-dev/taichi: docs/lang/articles/advanced/differentiable_programming.md

Taichi programs that violate this rule will result in an error.

:::danger DANGER
Violation of rules above might result in incorrect gradient result without a proper error.
We're actively working on improving the error reporting mechanism for it. Please feel free
to open a [github issue](https://github.com/taichi-dev/taichi/issues/new?assignees=&labels=potential+bug&template=bug_report.md&title=)
if you see any silent wrong results.

> :::

Reviewer: ```suggestion
:::

```

            ...
```

## Extending Taichi Autodiff system

> Sometimes user may want to override the gradients provided by the Taichi autodiff system. For example, when differentiating a 3D singular value decomposition (SVD) done with an iterative

Reviewer: ```suggestion
Sometimes user may want to override the gradients provided by the Taichi autodiff system. For example, when differentiating a 3D singular value decomposition (SVD) used in an iterative
```

solver, it is preferred to use a manually engineered SVD derivative subroutine for better numerical stability.
Taichi provides two decorators `ti.ad.grad_replaced` and `ti.ad.grad_for` to overwrite the default
automatic differentiation behavior.

> Here's a simple example to use customized gradient in autodiff:

Reviewer: how about calling it `customized gradient function` for a bit better clarity?
